deepmd/train/trainer_mt.py

```diff
@@ -0,0 +1,713 @@
+#!/usr/bin/env python3
+import logging
+import os
+import time
+import shutil
+import google.protobuf.message
+import numpy as np
+from deepmd.env import tf
+from deepmd.env import default_tf_session_config
+from deepmd.env import GLOBAL_TF_FLOAT_PRECISION
+from deepmd.env import GLOBAL_ENER_FLOAT_PRECISION
+from deepmd.fit import EnerFitting, WFCFitting, PolarFittingLocFrame, PolarFittingSeA, GlobalPolarFittingSeA, DipoleFittingSeA
+from deepmd.descriptor import DescrptLocFrame
+from deepmd.descriptor import DescrptSeA
+from deepmd.descriptor import DescrptSeT
+from deepmd.descriptor import DescrptSeAEbd
+from deepmd.descriptor import DescrptSeAEf
+from deepmd.descriptor import DescrptSeR
+from deepmd.descriptor import DescrptSeAR
+from deepmd.descriptor import DescrptHybrid
+from deepmd.model import EnerModel, WFCModel, DipoleModel, PolarModel, GlobalPolarModel
+from deepmd.loss import EnerStdLoss, EnerDipoleLoss, TensorLoss
+from deepmd.utils.errors import GraphTooLargeError
+from deepmd.utils.learning_rate import LearningRateExp
+from deepmd.utils.neighbor_stat import NeighborStat
+from deepmd.utils.sess import run_sess
+from deepmd.utils.type_embed import TypeEmbedNet
+
+from tensorflow.python.client import timeline
+from deepmd.env import op_module
+
+# load grad of force module
+import deepmd.op
+
+from deepmd.common import j_must_have, ClassArg
+
+log = logging.getLogger(__name__)
+
+
+def _is_subdir(path, directory):
+    path = os.path.realpath(path)
+    directory = os.path.realpath(directory)
+    if path == directory:
+        return False
+    relative = os.path.relpath(path, directory) + os.sep
+    return not relative.startswith(os.pardir + os.sep)
+
+def _generate_descrpt_from_param_dict(descrpt_param):
+    try:
+        descrpt_type = descrpt_param['type']
+    except KeyError:
+        raise KeyError('the type of descriptor should be set by `type`')
+    descrpt_param.pop('type', None)
+    to_pop = []
+    for kk in descrpt_param:
+        if kk[0] == '_':
+            to_pop.append(kk)
+    for kk in to_pop:
+        descrpt_param.pop(kk, None)
+    if descrpt_type == 'loc_frame':
+        descrpt = DescrptLocFrame(**descrpt_param)
+    elif descrpt_type == 'se_e2_a' or descrpt_type == 'se_a' :
+        descrpt = DescrptSeA(**descrpt_param)
+    elif descrpt_type == 'se_e2_r' or descrpt_type == 'se_r' :
+        descrpt = DescrptSeR(**descrpt_param)
+    elif descrpt_type == 'se_e3' or descrpt_type == 'se_at' or descrpt_type == 'se_a_3be' :
+        descrpt = DescrptSeT(**descrpt_param)
+    elif descrpt_type == 'se_a_tpe' or descrpt_type == 'se_a_ebd' :
+        descrpt = DescrptSeAEbd(**descrpt_param)
+    elif descrpt_type == 'se_a_ef' :
+        descrpt = DescrptSeAEf(**descrpt_param)
+    elif descrpt_type == 'se_ar' :
+        descrpt = DescrptSeAR(descrpt_param)
+    else :
+        raise RuntimeError('unknow model type ' + descrpt_type)
+    return descrpt
+    
+
+
+class DPTrainer_mt (object):
```

Comment:
I think this class can inherit `DPTrainer` and remove duplicate methods.

Reply:
And I still keep this advice: this class can inherit `DPTrainer` and remove duplicate methods. Or you can consider to create a new class `DPTrainerBase` to merge the same methods.